### PR TITLE
feat: Add corporate card validation

### PR DESCRIPTION
**Pull request by OmkarJ13**
[workerB](https://workerb.linearb.io/v2/badge/collaboration-page?magicLinkId=x95s1MV)
### <samp>🤖 Generated by Copilot at 46cc317</samp>

This pull request adds the feature of enrolling corporate cards from the platform APIs to the Fyle mobile app. It modifies the `RealTimeFeedService` and the `CorporateCreditCardExpenseService` to handle the API calls and responses. It also adds new interfaces for the API payload and response data structures in the `models/platform` folder. It updates the `AddCorporateCardComponent` and the `ManageCorporateCardsPage` to implement the form validation, error handling, and UI logic for the card enrollment feature. It also adds and modifies some style rules in the `scss` files for the card components.

### <samp>🤖 Generated by Copilot at 46cc317</samp>

> _Enroll the cards of doom, unleash the platform API_
> _Validate the data, show the errors in red_
> _Load and reload the cards, shimmer in the dark_
> _Pop the result, face the fate of the corporate_

### <samp>🤖 Generated by Copilot at 46cc317</samp>

*  Add new interfaces for the platform API payload, response, and error data structures ([link](https://github.com/fylein/fyle-mobile-app/pull/2250/files?diff=unified&w=0#diff-7e4c2935ca542c59b6fc0f52d94389ba72703c889068dd9e0a5d1add43074e68R1-R4), [link](https://github.com/fylein/fyle-mobile-app/pull/2250/files?diff=unified&w=0#diff-080a471281fd05cb224df701646ff03cb526d2344b6dac7d2e3248595e04a093R1-R5), [link](https://github.com/fylein/fyle-mobile-app/pull/2250/files?diff=unified&w=0#diff-e094ab5fc2d4fcc5f9b96a9d9f226ea38cd80595b68f9ad0572ba61df15541feR1-R6), [link](https://github.com/fylein/fyle-mobile-app/pull/2250/files?diff=unified&w=0#diff-414ee6f7356bccb89a277bbfc9850242f46c2b9e61b623d64d3a3c1b0619546dR1-R3))
*  Add new interface for the card enrollment payload data structure ([link](https://github.com/fylein/fyle-mobile-app/pull/2250/files?diff=unified&w=0#diff-7e4c2935ca542c59b6fc0f52d94389ba72703c889068dd9e0a5d1add43074e68R1-R4))
*  Remove the `@Cacheable()` decorator from the `getCorporateCards` method to avoid stale data ([link](https://github.com/fylein/fyle-mobile-app/pull/2250/files?diff=unified&w=0#diff-7b34d5803a4cd943ce1a77892b976fe94243cffde54b409e5ccf748500cfa220L42))
*  Inject the `SpenderPlatformV1ApiService` as a dependency in the `RealTimeFeedService` constructor ([link](https://github.com/fylein/fyle-mobile-app/pull/2250/files?diff=unified&w=0#diff-aea5e7bc6f6a6327858f8ff03ba55f605d40ebd49f410603a103e1e787bae292R16-R89))
*  Add methods to the `RealTimeFeedService` to validate, get the type, and enroll the card number using the platform API ([link](https://github.com/fylein/fyle-mobile-app/pull/2250/files?diff=unified&w=0#diff-aea5e7bc6f6a6327858f8ff03ba55f605d40ebd49f410603a103e1e787bae292R3-R10))
*  Add custom validators to the `cardForm` control in the `AddCorporateCardComponent` to check the validity and the issuer of the card number ([link](https://github.com/fylein/fyle-mobile-app/pull/2250/files?diff=unified&w=0#diff-85cca3686bdc76c7d1757fed2d3cfd9e2c68755b19b4d0edc64a6fab2b1c771dL33-R36), [link](https://github.com/fylein/fyle-mobile-app/pull/2250/files?diff=unified&w=0#diff-85cca3686bdc76c7d1757fed2d3cfd9e2c68755b19b4d0edc64a6fab2b1c771dR105-R147))
*  Add error handling and success handling logic to the `enrollCard` method in the `AddCorporateCardComponent` ([link](https://github.com/fylein/fyle-mobile-app/pull/2250/files?diff=unified&w=0#diff-85cca3686bdc76c7d1757fed2d3cfd9e2c68755b19b4d0edc64a6fab2b1c771dL50-R68), [link](https://github.com/fylein/fyle-mobile-app/pull/2250/files?diff=unified&w=0#diff-85cca3686bdc76c7d1757fed2d3cfd9e2c68755b19b4d0edc64a6fab2b1c771dR105-R147))
*  Add dynamic class and error messages to the card input element in the `add-corporate-card.component.html` file based on the `cardForm` state ([link](https://github.com/fylein/fyle-mobile-app/pull/2250/files?diff=unified&w=0#diff-4e011a42cc83184b53a781fc09dfbe3eed19d4830bffc7fde2e27757934cf4b6L20-R23), [link](https://github.com/fylein/fyle-mobile-app/pull/2250/files?diff=unified&w=0#diff-4e011a42cc83184b53a781fc09dfbe3eed19d4830bffc7fde2e27757934cf4b6L30-R83))
*  Add style rule for the error messages in the `add-corporate-card.component.scss` file ([link](https://github.com/fylein/fyle-mobile-app/pull/2250/files?diff=unified&w=0#diff-bfcd5ac8d936b76f67c888c3ab4d32d8f299658abe49190d2cf8a1bfd22e3bc4R80-R90))
*  Add shimmer placeholder and zero state for the cards list in the `manage-corporate-cards.page.html` file ([link](https://github.com/fylein/fyle-mobile-app/pull/2250/files?diff=unified&w=0#diff-a9990626fe6e417ac0088e9e79ea18f1e0b64e1f644e8cead61870540da85a5eL15-R31))
*  Add `BehaviorSubject` and boolean flag to the `ManageCorporateCardsPage` to trigger and indicate the card loading ([link](https://github.com/fylein/fyle-mobile-app/pull/2250/files?diff=unified&w=0#diff-450aa28d32fd44fd50bd49190bd3028ca78d9b54b8849b79192b7a2fc10a008fR27-R30))
*  Add logic to the `ionViewWillEnter` method in the `ManageCorporateCardsPage` to subscribe to the card loading observable ([link](https://github.com/fylein/fyle-mobile-app/pull/2250/files?diff=unified&w=0#diff-450aa28d32fd44fd50bd49190bd3028ca78d9b54b8849b79192b7a2fc10a008fL40-R53))
*  Add logic to the `addCorporateCard` method in the `ManageCorporateCardsPage` to handle the popover response and reload the cards if successful ([link](https://github.com/fylein/fyle-mobile-app/pull/2250/files?diff=unified&w=0#diff-450aa28d32fd44fd50bd49190bd3028ca78d9b54b8849b79192b7a2fc10a008fL138-R154))

## Clickup
https://app.clickup.com

## Code Coverage
Please add code coverage here

## UI Preview
https://www.loom.com/share/2f37755a080947dfa859ba1d1272e879

### CODE REVIEW

File: src/app/core/services/real-time-feed.service.ts [hunk 1]
```diff
@@ -1,10 +1,92 @@
...
+import { SpenderPlatformV1ApiService } from './spender-platform-v1-api.service';
+import { PlatformApiError } from '../models/platform/platform-api-error.model';
 
 @Injectable({
   providedIn: 'root',
 })
 export class RealTimeFeedService {
+  constructor(private spenderPlatformV1ApiService: SpenderPlatformV1ApiService) {}
+
+  isCardNumberValid(cardNumber: string): boolean {
```
Comment (OmkarJ13) [inline src/app/core/services/real-time-feed.service.ts:18]: Copied the same algorithm we use currently in web app

File: src/app/fyle/manage-corporate-cards/add-corporate-card/add-corporate-card.component.ts [hunk 2]
```diff
@@ -84,4 +109,47 @@ export class AddCorporateCardComponent implements OnInit {
...
+      return { invalidCardIssuer: true };
+    }
+
+    if (!this.isVisaRTFEnabled && cardType === RTFCardType.VISA) {
+      return { invalidCardIssuer: true };
+    }
+
+    if (!this.isMastercardRTFEnabled && cardType === RTFCardType.MASTERCARD) {
+      return { invalidCardIssuer: true };
+    }
```
Comment (Chethan-Fyle) [inline src/app/fyle/manage-corporate-cards/add-corporate-card/add-corporate-card.component.ts:142]:
```
if (
(!this.isYodleeEnabled && cardType === RTFCardType.OTHERS)
|| (!this.isVisaRTFEnabled && cardType === RTFCardType.VISA)
|| (!this.isMastercardRTFEnabled && cardType === RTFCardType.MASTERCARD)
) {
  return { invalidCardIssuer: true };
}
```

**Review by Chethan-Fyle — APPROVED** [review]
Minor comments.

File: src/app/fyle/manage-corporate-cards/add-corporate-card/add-corporate-card.component.ts (hunk 2 above)
Comment (OmkarJ13) [reply src/app/fyle/manage-corporate-cards/add-corporate-card/add-corporate-card.component.ts:133]: Fixed

File: src/app/fyle/manage-corporate-cards/manage-corporate-cards.page.ts [hunk 3]
```diff
@@ -37,7 +42,14 @@ export class ManageCorporateCardsPage {
   }
 
   ionViewWillEnter(): void {
-    this.corporateCards$ = this.corporateCreditCardExpenseService.getCorporateCards();
+    this.isLoadingCards = true;
+
+    this.corporateCards$ = this.loadCorporateCards$.pipe(
+      switchMap(() => this.corporateCreditCardExpenseService.getCorporateCards()),
+      tap(() => {
```
[the remark this answers is not in the document]
Comment (OmkarJ13) [reply src/app/fyle/manage-corporate-cards/manage-corporate-cards.page.ts]: I have removed the `isLoadingCards` variable here, it is not needed actually, To make it simpler, I have just used the async pipe and  `ngIf/else` to track the loading status.

**Review by Dimple16 — APPROVED** [review]

**Pull request merged**